### PR TITLE
Feature: workings 新增欄位

### PR DESCRIPTION
**Pull request by mark86092**

### CODE REVIEW

**Comment by mark86092** [comment]
`has_overtime_salary` will be updated

**Comment by mark86092** [comment]
Test fail

File: routes/workings.js [hunk 1]
```diff
@@ -274,22 +274,25 @@ function validateWorking(data) {
     }
 
     if (data.has_overtime_salary) {
-        if (data.has_overtime_salary === '0') {
-            data.has_overtime_salary = false;
-        } else if (data.has_overtime_salary === '1') {
-            data.has_overtime_salary = true;
-        } else {
-            throw new HttpError('has_overtime_salary should be 0 or 1', 422);
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) === -1){
```
Comment (mark86092) [inline routes/workings.js]: follow other coding style

```
if () {
```

File: routes/workings.js [hunk 2]
```diff
@@ -274,22 +274,25 @@ function validateWorking(data) {
...
+            throw new HttpError('加班是否有加班費應為是/否/不知道', 422);
+        }      
+    }
+
+    if(data.overtime_salary_is_legal && data.has_overtime_salary){
+        if(data.has_overtime_salary !== "yes"){
+            throw new HttpError('加班應有加班費，本欄位才有意義', 422);
+        }
+        else{
```
Comment (mark86092) [inline routes/workings.js]:
```
} else {
```

File: routes/workings.js [hunk 3]
```diff
@@ -274,22 +274,25 @@ function validateWorking(data) {
...
     }
 
     if (data.has_compensatory_dayoff) {
-        if (data.has_compensatory_dayoff === '0') {
-            data.has_compensatory_dayoff = false;
-        } else if (data.has_compensatory_dayoff === '1') {
-            data.has_compensatory_dayoff = true;
-        } else {
-            throw new HttpError('has_compensatory_dayoff should be 0 or 1', 422);
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) <= -1){
```
Comment (mark86092) [inline routes/workings.js]:
```
=== -1
```

as your other part code

File: routes/workings.js [hunk 4]
```diff
@@ -268,6 +272,29 @@ function validateWorking(data) {
...
+        }
+        else{
+            if(["yes", "no", "don't know"].indexOf(data.overtime_salary_is_legal) === -1){
+                throw new HttpError('加班費是否合法應為是/否/不知道', 422);
+            }
+        }
+    }
+
+    if (data.has_compensatory_dayoff) {
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) <= -1){
```
Comment (mark86092) [inline routes/workings.js]: === -1, as your other part code

File: routes/workings.js [hunk 5]
```diff
@@ -268,6 +272,29 @@ function validateWorking(data) {
             throw new HttpError("公司/單位名稱必填", 422);
         }
     }
+
+    if (data.has_overtime_salary) {
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) === -1){
```
Comment (mark86092) [inline routes/workings.js]: follow other coding style

```
if () {
```

File: test/testWorkings.js [hunk 6]
```diff
@@ -188,6 +188,115 @@ describe('Workings 工時資訊', function() {
...
+                        company_id: '00000001',
+                        has_overtime_salary: '-1',
+                    }))
+                    .expect(422)
+                    .end(done);
+            });
+        });
+
+        describe('overtime_salary_is_legal', function(){
+            for (let case of ['yes', 'no', 'don\'t know']) {
```
Comment (mark86092) [inline test/testWorkings.js]: it seems `case` is a preserved word

**Comment by mark86092** [comment]
BTW, can you add test to explain these fields are optional

File: routes/workings.js [hunk 7]
```diff
@@ -268,6 +272,28 @@ function validateWorking(data) {
...
         }
     }
+
+    if (data.has_overtime_salary) {
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) === -1) {
+            throw new HttpError('加班是否有加班費應為是/否/不知道', 422);
+        }      
+    }
+
+    if(data.overtime_salary_is_legal && data.has_overtime_salary) {
```
Comment (mark86092) [inline routes/workings.js]: what if the case `data.overtime_salary_is_legal === 'yes'`, and `data.has_overtime_salary` is undefined?

File: routes/workings.js [hunk 8]
```diff
@@ -268,6 +272,28 @@ function validateWorking(data) {
...
+            throw new HttpError('加班應有加班費，本欄位才有意義', 422);
+        } else {
+            if(["yes", "no", "don't know"].indexOf(data.overtime_salary_is_legal) === -1) {
+                throw new HttpError('加班費是否合法應為是/否/不知道', 422);
+            }
+        }
+    }
+
+    if (data.has_compensatory_dayoff) {
+        if(["yes", "no", "don't know"].indexOf(data.has_overtime_salary) === -1) {
```
Comment (mark86092) [inline routes/workings.js]: typo, `has_overtime_salary` --> `has_compensatory_dayoff`

Comment (mark86092) [inline routes/workings.js]: This is one of the reason test fail

File: routes/workings.js (hunk 7 above)
Comment (barry800414) [inline routes/workings.js]: L282應該就會要求兩個都是undefined了?

Comment (mark86092) [inline routes/workings.js]: L282 要求兩個都是undefined 沒錯，所以 `data.overtime_salary_is_legal === 'yes'`, and `data.has_overtime_salary` is undefined 怎麼辦？

Comment (barry800414) [inline routes/workings.js]: 不是，我講錯，兩個都要不是undefined，才會進去檢查。

Comment (mark86092) [inline routes/workings.js]: YES, 所以，這邊還是少掉我講的 case 沒處理～

**Comment by barry800414** [comment]
I have fixed all bugs and add tests, please reviewed it again.

File: test/testWorkings.js [hunk 9]
```diff
@@ -218,6 +218,20 @@ describe('Workings 工時資訊', function() {
                         .end(done);
                 });
             }
+            for (let input of ['', undefined]) { 
+                it('should not be uploaded to database if it is "' + input + '"', function(done) {
```
Comment (mark86092) [inline test/testWorkings.js]: How do you make sure it is not uploaded to db?

**Comment by mark86092** [comment]
現在有測試了這些 has_compensatory_dayoff 有值的情況，
{
company_id: '00000001',
has_compensatory_dayoff: '',
}
還要再麻煩你測試一個：（也就是 has_compensatory_dayoff 是 optional）
{
company_id: '00000001',
}

File: test/testWorkings.js (hunk 9 above)
Comment (barry800414) [inline test/testWorkings.js]: 我是用這個檢查 assert.notProperty() 有沒有那個欄位耶

Comment (mark86092) [inline test/testWorkings.js]: 沒錯，所以看不出來有沒有放到db裡，只知道回傳的response沒有這欄位，L222要寫的精確一點

**Comment by mark86092** [comment]
will you update the `overtime_salary_is_legal` field name to `is_overtime_salary_legal `

**Comment by barry800414** [comment]
我忘記這件事，但現在改好了!

**Pull request merged**